### DISCUSSION TITLE
Counting of duplicates in a wildcard list

### DISCUSSION THREAD

**Discussion by RayQuin** [adieyal/sd-dynamic-prompts#84] [General]
Hi, 

First of all, thanks for making this extension, it's great!
I've seen this issue which is somewhat similar to what I'm trying to do: https://github.com/adieyal/sd-dynamic-prompts/issues/57 

I have no coding knowledge so I'm not sure if I'm way off base but essentially, I'm trying to control probability of certain colours in a wildcards text file by having multiple identical copies of the same word in the list, but as far as I can tell, the dynamic prompt code is smart enough to account for duplicates, and so it isn't working the way I'd hoped.

For example, I have a list of colours like this:

brown
green
red 
yellow
blue
purple
pink

but I want most of the results to be brown or green, so I made a list like this:

brown
brown
brown
brown
brown
green
green
green
green
green
red 
yellow
blue
purple
pink

I hoped that 2/3rds of the results would be brown or green, but it isn't hitting that ratio.

I've tried the Jinji script and wrote it all out in that, but then when I activated the Jinji script button in the dropdown, it made the other wildcard calls in the prompt not work. Is there a prefix or other way to make it call wildcard text files while in Jinji mode without having to have all the lists be formatted in the Jinji code, as some of the wildcards lists are really long?

or is there a way I can edit it to make your script not account for if there are duplicates in the list?

Really sorry if I've not explained this, and it's not a big deal if it won't work.

**Comment by adieyal** [comment]
Apologies for only responding now - I completely missed this. Currently, the script automatically de-duplicates wildcards files. There has been another request to manipulate the probability of a wildcard being chosen. Could you please explain your use case? I would like to understand whether it is general enough to implement it as a feature.

**Reply by RayQuin** [reply]
Thanks for getting in touch - don't worry about the delay at all.


my use case was just because I'm using a few wildcard text files all to combine to make portraits of rpg characters, so I'd have "__haircolour__ __hairstyle__, wearing __armourcolour__ __armourstyle__ on a __backgroundcolour__ background", 
but I was hoping that for the colours to have mostly 'normal' colours like ginger hair, brown hair, and mostly 'steel' or 'brown leather' armour, while still having some interesting colours.
 I was essentially finding that if I had a less normal colour hair, like say "Ice-blue hair", and a less normal colour choice like "amethyst purple" for say wizard robes, that the colours would only rarely work, whereas if I had an 'odd' colour for hair, but a more conventional colour or material for the clothes, or vice versa, it would be much more likely that SD would get the odd colour right (and it usually made a more striking picture!). So I wanted to control probability so that if I'd randomly gotten an odd colour for one element of the prompt, that I'd be more likely to have a more normal one for the other parts of the prompt. 



I've actually solved my issue by using the ability to use the asterisk at the end to call from multiple numbered text files with different colours in each,
so as an example I now just use __haircolour*__
and have 5 text files, with the odd colours all in haircolour1.txt, and just brunette, blond, ginger, grey and black in haircolours2-5.txt, so that it is more likely to pick more normal hair colours, and I've done the same with the armourcolour wildcard text file.

**Reply by ZombieAlex** [reply]
For what it's worth, I just stumbled on the same use-case and found this discussion via search. My inline prompt was something like `{black|brown|0.2::blonde|0.2::colorful} hair` and as my overall prompt length started to get very large I began migrating to wildcard files but it threw off my ratios as Ray noticed.

I can work around it in the same way that Ray did, but I think it'd be better if the code either:
1. Didn't de-duplicate the lists
2. Or supported the inline weighting syntax in wildcard files. So, for instance, I could have a single line in the wildcard file like `0.2::blonde` to achieve my lowered blonde weighting from the inline prompt.

Cheers!
